hivemind/client/averaging/matchmaking.py

```diff
@@ -0,0 +1,367 @@
+            if maybe_next_leader is None or maybe_next_leader == self.endpoint:
+                continue  # do not count myself as a potential leader for me (even at a later expiration time)
+
+            maybe_group_allreduce = await self.request_join_group(maybe_next_leader, self.declared_expiration_time)
+            if maybe_group_allreduce is not None:
+                return maybe_group_allreduce
+
+    async def request_join_group(self, leader: Endpoint, expiration_time: DHTExpiration) -> Optional[GroupAllReduce]:
+        """
+        :param leader: request this peer to be your leader for allreduce
+        :param expiration_time: inform leader that we intend to begin averaging before this expiration_time
+        :returns: if leader leader accepted us and started AllReduce, return that AllReduce. Otherwise, return None
+        """
+        assert self.looking_for_group and self.current_leader is None
+        call: Optional[grpc.aio.UnaryStreamCall[averaging_pb2.JoinRequest, averaging_pb2.MessageFromLeader]] = None
+        try:
+            async with self.lock_request_join_group:
+                leader_stub = ChannelCache.get_stub(leader, averaging_pb2_grpc.DecentralizedAveragingStub, aio=True)
+                call = leader_stub.rpc_join_group(averaging_pb2.JoinRequest(
+                    endpoint=self.endpoint, schema_hash=self.schema_hash, expiration=expiration_time))
+
+                message = await call.read()  # TODO use timeout?
+                if message.code != averaging_pb2.ACCEPTED:
+                    code = averaging_pb2.MessageCode.Name(message.code)
+                    logger.debug(f"{self.endpoint} - requested {leader} to be my leader, but got rejected with {code}")
+                    return None
+
+                # else: we were accepted
+                logger.debug(f"{self.endpoint} - joining the group of {leader}; waiting for peers")
+                self.current_leader = leader
+                if len(self.current_followers) > 0:
+                    await self.leader_disband_group()
+
+            message = await call.read()
+            if message.code == averaging_pb2.BEGIN_ALLREDUCE:
+                async with self.lock_request_join_group:
+                    return await self.follower_assemble_group(leader, message.group_id, message.ordered_group_endpoints)
+            elif message.code == averaging_pb2.GROUP_DISBANDED and bool(message.suggested_leader):
+                logger.debug(f"{self} - leader disbanded group and redirected us to {message.suggested_leader}")
+                return await self.request_join_group(message.suggested_leader, expiration_time)
+
+            else:
+                logger.debug(f"{self} - leader sent {averaging_pb2.MessageCode.Name(message.code)}, leaving group")
+                return None
+        finally:
+            self.current_leader = None
+            if call is not None:
+                call.cancel()
```

Comment:
```

        async with self.lock_request_join_group:
            leader_stub = ChannelCache.get_stub(leader, averaging_pb2_grpc.DecentralizedAveragingStub, aio=True)
            call = leader_stub.rpc_join_group(averaging_pb2.JoinRequest(
                endpoint=self.endpoint, schema_hash=self.schema_hash, expiration=expiration_time))
            try:
                message = await call.read()  # TODO use timeout?
                if message.code != averaging_pb2.ACCEPTED:
                    code = averaging_pb2.MessageCode.Name(message.code)
                    logger.debug(f"{self.endpoint} - requested {leader} to be my leader, but got rejected with {code}")
                    return None

                # else: we were accepted
                logger.debug(f"{self.endpoint} - joining the group of {leader}; waiting for peers")
                self.current_leader = leader
                if len(self.current_followers) > 0:
                    await self.leader_disband_group()

                message = await call.read()
                if message.code == averaging_pb2.BEGIN_ALLREDUCE:
                    async with self.lock_request_join_group:
                        return await self.follower_assemble_group(leader, message.group_id, message.ordered_group_endpoints)
                elif message.code == averaging_pb2.GROUP_DISBANDED and bool(message.suggested_leader):
                    logger.debug(f"{self} - leader disbanded group and redirected us to {message.suggested_leader}")
                    return await self.request_join_group(message.suggested_leader, expiration_time)

                else:
                    logger.debug(f"{self} - leader sent {averaging_pb2.MessageCode.Name(message.code)}, leaving group")
                    return None
            finally:
                self.current_leader = None
                if call is not None:
                    call.cancel()
```

Reply:
found an issue with this implementation: it acquires lock_request_join_group for longer than required.
As a result, it paralyzes all incoming requests and itself (since it tries to acquire the lock inside a context that already acquired that lock).